src/loader.js

```diff
@@ -0,0 +1,91 @@
+let assume = require('assume');
+let debug = require('debug')('taskcluster:loader');
+let Promise = require('promise');
+
+function loader(componentDirectory, requiredVirtualComponents) {
+  /**
+   * If present, this is a list of component names that a loader
+   * must define.  These are likely things like configuration values
+   */
+  if (requiredVirtualComponents) {
+    assume(requiredVirtualComponents).is.array();
+  } else {
+    requiredVirtualComponents = [];
+  }
+
+  return function(virtualComponents) {
+    let initializedComponents = {};
+
+    function loadComponent(name, visited) {
+      assume(name).to.be.ok();
+
+      if (!visited) {
```

Comment:
Okay, I see what you're doing... I don't like how you give promises as arguments to the functions...
With the upfront cycle detection this could really be as simple as:
```js
function loader(componentDirectory, requiredVirtualComponents) {
  // Sanity checks of arguments and ensure it's a DAG... (just copy in from snippets above)
  let components = requiredVirtualComponents.concat(_.keys(componentDirectory));
  return function(virtualComponents, component) {
    assert(components.includes(component), "component not defined");
    var initilizing = {};
    for(let c of requiredVirtualComponents) {
      assert(virtualComponents[c] === undefined, c + " is required");
      initilizing[c] = Promise.resolve(virtualComponents[c]);
    }
    
    function init(name) {
      if (!initilizing[name]) {
        var def = componentDirectory[name];
        initilizing[name] = Promise.all(def.requires.map(c => {
          return init(c);
        })).then(inputs => {
          var ctx = {};
          assert(def.requires.length === inputs.length, "internal error");
          for(let i = 0; i < inputs.length; i++) {
            ctx[def.requires[i]] = inputs[i];
          }
          return def.setup.call(null, ctx);
        });
      }
      return initilizing[name];
    };
    
    return init(component);
  };
}
```

I might suggest renaming `requiredVirtualComponents` to `requiredInputs` and then `virtualComponents` to `inputs`.. it feels weird to call it components... Anyways, this an approach...